core/store/src/trie/update.rs

```diff
@@ -2,19 +2,51 @@ use std::collections::{BTreeMap, HashMap, HashSet};
 use std::iter::Peekable;
 use std::sync::Arc;
 
+use borsh::{BorshDeserialize, BorshSerialize};
+
 use near_primitives::hash::CryptoHash;
 
 use crate::trie::TrieChanges;
+use crate::StorageError;
 
 use super::{Trie, TrieIterator};
-use crate::StorageError;
+
+/// A structure used to index state changes due to transaction/receipt processing and other things.
+#[derive(BorshSerialize, BorshDeserialize, Clone)]
+pub enum KVChangeCause {
+    /// A type of update that does not get finalized. Used for verification and execution of
+    /// immutable smart contract methods. Attempt fo finalize a `TrieUpdate` containing such
+    /// change will lead to panic.
+    NonFinalizable,
+    /// A type of update that is used to mark the initial storage update, e.g. during genesis
+    /// or in tests setup.
+    InitialState,
```

Comment:
Please, review if there is a better way to express this intent or completely avoid it somehow.

Reply:
I cannot think of anything better.